src/blas/KokkosBlas3_trsm.hpp

```diff
@@ -0,0 +1,184 @@
+                     (trans[0] == 'T') || (trans[0] == 't')||
+                     (trans[0] == 'C') || (trans[0] == 'c');
+  bool valid_diag  = (diag[0] == 'U' ) || (diag[0] == 'u' )||
+                     (diag[0] == 'N' ) || (diag[0] == 'n' );
+  if(!valid_side) {
+    std::ostringstream os;
+    os << "KokkosBlas::trsm: side = '" << side[0] << "'. " <<
+      "Valid values include 'L' or 'l' (A is on the left of X), "
+      "'R' or 'r' (A is on the right of X).";
+    Kokkos::Impl::throw_runtime_exception (os.str ());
+  }
+  if(!valid_uplo) {
+    std::ostringstream os;
+    os << "KokkosBlas::trsm: uplo = '" << uplo[0] << "'. " <<
+      "Valid values include 'U' or 'u' (A is upper triangular), "
+      "'L' or 'l' (A is lower triangular).";
+    Kokkos::Impl::throw_runtime_exception (os.str ());
+  }
+  if(!valid_trans) {
+    std::ostringstream os;
+    os << "KokkosBlas::trsm: trans = '" << trans[0] << "'. " <<
+      "Valid values include 'N' or 'n' (No transpose), 'T' or 't' (Transpose), "
+      "and 'C' or 'c' (Conjugate transpose).";
+    Kokkos::Impl::throw_runtime_exception (os.str ());
+  }
+  if(!valid_diag) {
+    std::ostringstream os;
+    os << "KokkosBlas::trsm: diag = '" << diag[0] << "'. " <<
+      "Valid values include 'U' or 'u' (the diagonal of A is assumed to be unit), "
+     "'N' or 'n' (the diagonal of A is assumed to be non-unit).";
+    Kokkos::Impl::throw_runtime_exception (os.str ());
+  }
+
+  // Check compatibility of dimensions at run time.
+  bool A_s =  (side[0] == 'L' || side[0] == 'l');
+
+  int64_t A0 = A.extent(0);
+  int64_t A1 = A.extent(1);
+  int64_t B0 = B.extent(0);
+  int64_t B1 = B.extent(1);
+
+  if ((A0 != A1) || ((A_s?B0:B1) != A1)) {
+    std::ostringstream os;
+    os << "KokkosBlas::trsm: Dimensions of A and B do not match: "
+       << "side: " << side[0]
+       << " A: " << A.extent(0) << " x " << A.extent(1)
+       << " B: " << B.extent(0) << " x " << B.extent(1);
+    Kokkos::Impl::throw_runtime_exception (os.str ());
+  }
+
+  // Return if degenerated matrices are provided
+  if((A.extent(0) == 0) || (A.extent(1) == 0) || (B.extent(0) == 0) || (B.extent(1) == 0))
+    return;
+
+  // Minimize the number of Impl::TRSM instantiations, by
+  // standardizing on particular View specializations for its template
+  // parameters.
+  using AVT = Kokkos::View<typename AViewType::const_value_type**,
+                           typename AViewType::array_layout,
+                           typename AViewType::device_type,
+                           Kokkos::MemoryTraits<Kokkos::Unmanaged> >;
+  using BVT = Kokkos::View<typename BViewType::non_const_value_type**,
+                           typename BViewType::array_layout,
+                           typename BViewType::device_type,
+                           Kokkos::MemoryTraits<Kokkos::Unmanaged> >;
+
+  AVT A_i = A;
+  
+  if (BViewType::rank == 1) {
+    auto B_i = BVT(B.data(), B.extent(0), 1);
```

Comment:
Is the intent here to create a subview?  I'm not a huge fan of creating a View directly from a raw pointer, because it misses out on `Kokkos::subview`'s debug checking.  I see _why_ you want to do this -- you are trying to cover the `rank==1` and `rank==2` cases in a single code, without the benefit of C++17 `if constexpr`.  However, if `rank==1`, it would be better to call `TRSV`, which is optimized for the single-vector case.

Reply:
In summary, since you can't use `if constexpr` yet, I think you might be stuck with creating nonowning Views from raw pointers.  However, the single-vector case should really call `trsv`, not `trsm`.

Reply:
Thanks, @mhoemmen I understand your point. I just wanted to cover both rank==1 and rank==2 while we have not had ```TRSV``` yet. 
In the new update, only rank=2 is used and I will add ```TRSV``` interace in another PR.

Reply:
It would have been better to have started with TRSV and used it to build a first-pass implementation of TRSM, but OK, I understand.